net/freeradius/src/opnsense/scripts/Freeradius/generate_certs.php

```diff
@@ -68,14 +68,32 @@
                     // generate ca pem file
                     if (!empty($cert->caref)) {
                         $cert = (array)$cert;
-                        $ca_pem_content .= ca_chain($cert);
+                        $cert_pem_content .= ca_chain($cert);
                     }
                 }
             }
         }
 
+        $cert_refid = (string)$find_cert->ca;
+        // if eap has a ca-certificate attached, search for its contents
+        if ($cert_refid != "") {
+            foreach ($configObj->ca as $ca) {
+                if ($cert_refid == (string)$ca->refid) {
+                    // generate cert pem file
+                    $pem_content = trim(str_replace("\n\n", "\n", str_replace(
+                        "\r",
+                        "",
+                        base64_decode((string)$ca->crt)
+                    )));
+
+                    $pem_content .= "\n";
+                    $cert_pem_content .= $pem_content;
+                }
```

Comment:
Shouldn't we call ca_chain($ca) as well? if if the CA is an intermediate it's going to be missing. Or, rather, do we even need the ca if we complete the chain cert like you suggested above? It looks like we're just going to append the same files again, aren't we?

Reply:
No, you are not going to append the same files again. The `cert` is the certificate the server uses to authenticate itself to the clients. It is the server certificate for TLS. The `root` certificate is the root ca for the clients. So the server knows that only clients with a cert signed by this root ca are allowed to connect. If you add the whole chain then client certificates can be signed by ANY of the ca's in the chain, see also <https://networkradius.com/doc/3.0.10/raddb/tls/tls-config_tls-common.html>:

>Trusted Root CA list: ALL of the CA’s in this list will be trusted to issue client certificates for authentication. In general, self-signed certificates should be used for 802.1x (EAP) authentication, and this CA file should contain ONLY one CA certificate.

Fixed using the wrong variable.

Reply:
Thanks for explaining. Terminology used in RADIUS seems confusing :/

Reply:
It is...